File: rcc/utils/single_instance.py
import os
import psutil

from rcc.utils.constants import LOCK_FILE
from framework.singleton import singleton


@singleton
class SingleInstance:
    """Application single instance handler"""

    def acquire(self, name) -> None:
        """Try to acquire lock file"""
        if os.path.exists(LOCK_FILE):
            with open(LOCK_FILE, "r") as f:
                pid = int(f.read().strip())

            if psutil.pid_exists(pid) and psutil.Process(pid).name().startswith(name):
                print(f"Application already running with pid {pid}, killing...")
                psutil.Process(pid).kill()
            else:
                os.remove(LOCK_FILE)

        with open(LOCK_FILE, "w") as f:
            f.write(str(os.getpid()))


SINGLE_INSTANCE = SingleInstance()

File: RogControlCenter/src/utils/single_instance.cpp
#include "../../include/utils/single_instance.hpp"

#include <csignal>
#include <filesystem>
#include <fstream>
#include <iostream>

#include "../../include/utils/constants.hpp"
#include "../../include/utils/string_utils.hpp"

void SingleInstance::acquire() {
	namespace fs = std::filesystem;

	if (fs::exists(Constants::LOCK_FILE)) {
		std::ifstream f(Constants::LOCK_FILE);
		pid_t pid;
		f >> pid;
		f.close();

		if (pid > 0 && kill(pid, 0) == 0) {
			std::cout << "Application already running with pid " + StringUtils::trim(std::to_string(pid)) + ", killing..." << std::endl;
			kill(pid, SIGKILL);
		} else {
			fs::remove(Constants::LOCK_FILE);
		}
	}

	std::ofstream out(Constants::LOCK_FILE);
	out << Constants::PID;
	out.close();
}
